### PR TITLE
Feature/setup tests

### PR DESCRIPTION
**Pull request by AP6YC**
This feature implements unit tests comparing incremental and batch variant equivalence. It also finalizes the CH implementation with an API and documentation that other implementations will soon follow.

### CODE REVIEW

File: src/cvi/modules/CH.py [hunk 1]
```diff
@@ -1,97 +1,126 @@
...
         else:
             n_new = self.n[i_label] + 1
-            v_new = (1 - 1/n_new) * self.v[:, i_label] + (1/n_new) * sample
-            delta_v = self.v[:, i_label] - v_new
+            v_new = (1 - 1/n_new) * self.v[i_label, :] + (1/n_new) * sample
+            delta_v = self.v[i_label, :] - v_new
             diff_x_v = sample - v_new
-            CP_new = self.P[i_label] + transpose(diff_x_v[np.newaxis])
-            G_new = self.G[:, i_label] + diff_x_v + self.n[i_label] * delta_v
+            CP_new = self.CP[i_label] + np.inner(diff_x_v, diff_x_v) + self.n[i_label]*np.inner(delta_v, delta_v) + 2*np.inner(delta_v,self.G[i_label, :])
```
Comment (nikmelton) [inline src/cvi/modules/CH.py:74]: What's goin on here? this looks computationally taxing

File: src/cvi/modules/common.py [hunk 2]
```diff
@@ -91,3 +95,34 @@ def setup(self, sample):
     #     # self.BGSS = np.single()
     #     # self.WGSS = np.single()
     #     # self.n_clusters = np.intc()
+
+# This decorator appends the docstring of one function to another
```
Comment (nikmelton) [inline src/cvi/modules/common.py:99]: cool

File: tests/test_cvi.py [hunk 3]
```diff
@@ -11,41 +11,206 @@
...
+    return sample, label
+
 # --------------------------------------------------------------------------- #
 # TESTS
 # --------------------------------------------------------------------------- #
 
 class TestCVI:
 
-    def test_opts(self):
+    def test_load_data(self, data) -> None:
```
Comment (nikmelton) [inline tests/test_cvi.py:133]: need typing on data

File: tests/test_cvi.py [hunk 4]
```diff
@@ -11,41 +11,206 @@
...
         sample = np.ones(3)
         my_cvi.setup(sample)
-        lg.info("asdf")
+
+        lg.info("After setup")
         lg.info(my_cvi)
-        return
\ No newline at end of file
+
+        return
```
Comment (nikmelton) [inline tests/test_cvi.py:216]: In the future lets add a test to not only verify the batch and incremental versions have the same results but also that batch is slower than the incremental version

File: tests/test_cvi.py [hunk 5]
```diff
@@ -11,41 +11,206 @@
...
+        },
+    }
+
+    return data_dict
+
+# --------------------------------------------------------------------------- #
+# UTILITY FUNCTIONS
+# --------------------------------------------------------------------------- #
+
+def get_cvis() -> list:
```
Comment (nikmelton) [inline tests/test_cvi.py:94]: use "from Typing import List" and the define the return type with -> List[cvi]

File: tests/test_cvi.py [hunk 6]
```diff
@@ -11,41 +11,206 @@
...
 import src.cvi as cvi
 # import ..
 
+# --------------------------------------------------------------------------- #
+# FIXTURES
+# --------------------------------------------------------------------------- #
+
+# Set the fixture scope to the testing session to load the data once
+@pytest.fixture(scope="session")
+def data() -> dict:
```
Comment (nikmelton) [inline tests/test_cvi.py:38]: use "from Typing import Dict" and the define the return type with -> Dict[str,Dict[str,Dict[str,np.ndarray]]]

File: src/cvi/modules/common.py [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import numpy as np
+from typing import Callable
 
 class LabelMap():
```
Comment (nikmelton) [inline src/cvi/modules/common.py:4]: maybe add a doc string here

File: src/cvi/modules/CH.py [hunk 8]
```diff
@@ -1,97 +1,126 @@
...
-    # # References
-    # 1. L. E. Brito da Silva, N. M. Melton, and D. C. Wunsch II, "Incremental Cluster Validity Indices for Hard Partitions: Extensions  and  Comparative Study," ArXiv  e-prints, Feb 2019, arXiv:1902.06711v1 [cs.LG].
-    # 2. T. Calinski and J. Harabasz, "A dendrite method for cluster analysis," Communications in Statistics, vol. 3, no. 1, pp. 1–27, 1974.
-    # 3. M. Moshtaghi, J. C. Bezdek, S. M. Erfani, C. Leckie, and J. Bailey, "Online Cluster Validity Indices for Streaming Data," ArXiv e-prints, 2018, arXiv:1801.02937v1 [stat.ML]. [Online].
-    # 4. M. Moshtaghi, J. C. Bezdek, S. M. Erfani, C. Leckie, J. Bailey, "Online cluster validity indices for performance monitoring of streaming data clustering," Int. J. Intell. Syst., pp. 1-23, 2018.
-    # """
-    # Calinski-Harabasz (CH) Cluster Validity Index.
-    # """
-
     def __init__(self):
```
Comment (nikmelton) [inline src/cvi/modules/CH.py:23]: add return type -> None

File: src/cvi/__init__.py [hunk 9]
```diff
@@ -14,7 +14,6 @@
 from .icvi import *
 from .modules import *
 
-
 class CVIOpts():
```
Comment (nikmelton) [inline src/cvi/__init__.py:17]: add doc string

**Review by nikmelton — CHANGES_REQUESTED** [review]
See comments. Overall looks good just a little house-keeping needed

File: src/cvi/modules/CH.py [hunk 10]
```diff
@@ -1,97 +1,126 @@
...
         if not self.mu.any():
             mu_new = sample
             self.setup(sample)
         else:
             mu_new = (1 - 1/n_samples_new) * self.mu + (1/n_samples_new) * sample
 
-        if i_label > self.n_clusters:
+        # IF NEW CLUSTER LABEL
+        # Correct for python 0-indexing
+        if i_label + 1 > self.n_clusters:
```
[the remark this answers is not in the document]
Comment (AP6YC) [reply src/cvi/modules/CH.py:53]: I was debating between the two myself too. I agree; I'll change it back to your suggestion.

File: tests/test_cvi.py (hunk 3 above)
Comment (AP6YC) [reply tests/test_cvi.py:133]: Good catch 👍

File: tests/test_cvi.py (hunk 4 above)
Comment (AP6YC) [reply tests/test_cvi.py:216]: That would be a good test. I'll make that an issue and new feature.

Comment (AP6YC) [reply tests/test_cvi.py:216]: Issue #5

File: tests/test_cvi.py (hunk 5 above)
Comment (AP6YC) [reply tests/test_cvi.py:94]: While I like that idea, I was wondering: would this typing fail for inheriting classes? All my CVI implementations inherit from a base CVI type, but does that mean that a linter would get upset that they are not strictly speaking CVIs?

File: tests/test_cvi.py (hunk 6 above)
Comment (AP6YC) [reply tests/test_cvi.py:38]: While that would be more specific, I think that I might just make a new TestData container to standardize things. I'll update the signature typing accordingly.

File: src/cvi/modules/common.py (hunk 7 above)
Comment (AP6YC) [reply src/cvi/modules/common.py:4]: Good call; I forgot to docstring some modules/files.

File: src/cvi/__init__.py (hunk 9 above)
Comment (AP6YC) [reply src/cvi/__init__.py:17]: This reminds me that I don't think that a CVI options class will be necessary. I'll remove it for now and make sure everything else has docstrings.

File: src/cvi/modules/CH.py (hunk 1 above)
Comment (AP6YC) [reply src/cvi/modules/CH.py:74]: This is the equivalent of your CP_update in icvi.py for CH. It has three vector dot products (np.inner is the same as np.dot for 1-D vectors, though I might change it to explicitly np.dot).

**Pull request merged**